include/seastar/rpc/rpc.hh

```diff
@@ -390,28 +392,63 @@ public:
 
 namespace internal {
 
-struct deferred_snd_buf {
-    promise<> pr;
-    snd_buf data;
+class snd_buf_batched_queue {
+    using vector_type = boost::container::small_vector<snd_buf*, 1>;
+
+    std::function<future<>(snd_buf*)> _process_func;
+    shard_id _processing_shard;
+    vector_type _queue;
+    vector_type _cur_batch;
+    vector_type::iterator _cur_batch_pos;
+    future<> _process_fut = make_ready_future();
+
+public:
+    snd_buf_batched_queue(std::function<future<>(snd_buf*)> process_func, shard_id processing_shard)
+        : _process_func(std::move(process_func))
+        , _processing_shard(processing_shard)
+    {}
+
+    void enqueue(snd_buf* buf);
+
+    future<> process_loop();
+};
+
+// Safely delete the original allocation buffer on the local shard
+// When deleted after it was sent on the remote shard, we queue
+// up the buffer pointers to be destroyed and deleted as a batch
+// back on the local shard.
+class snd_buf_deleter_impl final : public deleter::impl {
+    snd_buf* _obj_ptr;
+    snd_buf_batched_queue& _delete_queue;
+
+public:
+    snd_buf_deleter_impl(snd_buf* obj_ptr, snd_buf_batched_queue& delete_queue)
+        : impl(deleter())
+        , _obj_ptr(obj_ptr)
+        , _delete_queue(delete_queue)
+    {}
+
+    virtual ~snd_buf_deleter_impl() override {
+        _delete_queue.enqueue(_obj_ptr);
```

Comment:
This may throw, no?

Reply:
@avikivity what do you think regarding using intrusive lists for the queues vs. a vector?
Intrusive lists save allocations and make the buffers more easily transferable across shards.